File: Python3/1001 - 1200/1008 Construct Binary Search Tree from Preorder Traversal.py
# 40ms, 100.0%; 13.1MB, 100.0%
class Solution:
    def bstFromPreorder(self, preorder: list) -> TreeNode:
        head = TreeNode(preorder[0])
        for i in range(1, len(preorder)):
            self.__insert(head, preorder[i])
        return head

    def __insert(self, root: TreeNode, x: int) -> None:
        while True:
            if x < root.val:
                if root.left:
                    root = root.left
                else:
                    root.left = TreeNode(x)
                    return
            elif root.right:
                root = root.right
            else:
                root.right = TreeNode(x)
                return

File: 1001 - 1200/1008 Construct Binary Search Tree from Preorder Traversal.cc
// 4ms, 100.0%; 9MB, 100.0%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    TreeNode *bstFromPreorder(const vector<int> &preorder) {
        TreeNode *head = new TreeNode(preorder[0]);
        for (int i = 1, size = preorder.size(); i < size; ++i) insert(head, preorder[i]);
        return head;
    }

private:
    void insert(TreeNode *root, int val) {
        while (true) {
            if (val < root->val) {
                if (root->left) {
                    root = root->left;
                } else {
                    root->left = new TreeNode(val);
                    return;
                }
            } else if (root->right) {
                root = root->right;
            } else {
                root->right = new TreeNode(val);
                return;
            }
        }
    }
};
